### ISSUE TITLE
New bug class with no scanner: guard-then-app-callback-then-deref (TOCTOU through interp2app dispatch)

### ISSUE THREAD

**Issue by devdanzin** [BHUVANSH855/pypy-review-toolkit#10]
`git-history-analyzer` surfaced a bug shape no scanner in the toolkit looks for, and it reproduces as a SIGSEGV on 4 methods. Filing it as a scanner proposal because the existing guard scanners **structurally cannot** find it — they check whether a guard is *present*, and here it is present and then invalidated.

## The shape

```python
def truncate_w(self, space, w_pos=None):
    self._check_attached(space)                          # guard PASSES
    space.call_method(self, "flush")                     # app-level dispatch -- subclass runs here
    return space.call_method(self.w_buffer, "truncate", w_pos)   # w_buffer is now None
```

`space.call_method(self, ...)` is an **app-level** method lookup. A subclass overriding `flush` runs arbitrary Python between the check and the use — and if that override calls `self.detach()`, the field the guard vouched for is `None` by the time it is dereferenced.

```python
import io
class T(io.TextIOWrapper):
    _d = False
    def flush(self):
        if not self._d:
            self._d = True
            self.detach()
t = T(io.BytesIO(b""), encoding="utf-8")
t.truncate()          # SIGSEGV
```

Reproduced on PyPy 7.3.23 (`194f9f44b505`):

| method | PyPy | CPython 3.14.4 |
|---|---|---|
| `truncate()` | **SIGSEGV** | **SIGSEGV** |
| `tell()` | **SIGSEGV** | **SIGSEGV** |
| `seek(0)` | **SIGSEGV** | **SIGSEGV** |
| `close()` | **SIGSEGV** | **SIGSEGV** |
| `write('x')` | ok | **SIGSEGV** |
| `read()` | ok | **SIGSEGV** |

Sites in `interp_textio.py`: `:827` `:838` `:1141` `:1244` `:1321`, plus `_read_chunk:894` and `_set_encoder_decoder:651`. Shared with CPython — `main` is immunized by `db4b1948bc4` (gh-143008), 3.13/3.14 are not, and PyPy has ported neither that nor gh-142594.

## Why no existing scanner sees it

- `scan_sibling_guard_consistency` — the method **calls** `self._check_attached(space)`. Passes.
- `scan_cross_class_method_guard` — a guard is present. Passes.
- Both answer "is there a guard?"; the question here is "does the guard still hold at the point of use?"

`W_TextIOWrapper.reconfigure` is the instructive contrast: it has *no* guard and is protected only by the accident of that same `flush` dispatch, so the sibling scanner does catch it. Same dispatch, opposite reason — which is what makes this a distinct class rather than a variant.

## Proposed check

Within one app-exposed method, flag the ordered triple:

1. a guard — `self._check_*(...)`, or `if self.<field>` — establishing `<field>`, then
2. a call that can re-enter app-level Python: `space.call_method(self, ...)`, `space.call_function(...)`, `space.getattr(...)`, `space.next(...)`, anything dispatching on a user-overridable object, then
3. a dereference of `<field>` — `space.call_method(self.<field>, ...)` or an attribute access on it.

Steps 1 and 3 are already implemented in the two guard scanners; the new part is step 2 and the ordering constraint.

### Calibration notes, from the sites that exist

- **`space.call_method(self, ...)` is the high-signal form** — dispatch on `self` is what a subclass overrides. Dispatch on an *internal* object (`self.w_decoder`) is weaker; probably worth a lower rank, not a suppression.
- **Do not credit `space.call_method(self, ...)` as a guard.** This is the trap I flagged when the FP-reduction work landed: if the delegation-recognition logic in ecb8434/f6c0a05 ever grows to treat app-level self-dispatch as a guard call, it will suppress `reconfigure` (a confirmed SIGSEGV) as well as miss this class. Currently correct — worth a regression test either way.
- **Re-checking after the dispatch is the fix**, so a second `_check_*` call between steps 2 and 3 should clear the finding. That is exactly what CPython's `db4b1948bc4` did: route every `self->buffer` access through a checking helper.
- Expect the same shape outside `_io` wherever an interp-level method dispatches back into app-level Python mid-body. `_io` is where it was found, not necessarily where it is worst.

Related: #1 records this as `PYPY-REV-008`. I have not written this scanner — filing the class, the reproducers, and the calibration notes so the design is on record. Happy to implement it if you want it, though I would rather you weigh in on the ranking heuristics first, since that is where this kind of check lives or dies.
